# Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# Import Population Data

In [2]:
path = r'C:\Users\Eric\Desktop\Data Analytics\Data Immersion\Achievement 6'

In [3]:
pop = pd.read_csv(os.path.join(path,'Data','Original Data','PopulationEstimates.csv'))

In [4]:
pop.head(15)

FIPStxt State        Area_Name  Rural-urban_Continuum Code_2003  \
0         0    US    United States                              NaN   
1      1000    AL          Alabama                              NaN   
2      1001    AL   Autauga County                              2.0   
3      1003    AL   Baldwin County                              4.0   
4      1005    AL   Barbour County                              6.0   
5      1007    AL      Bibb County                              1.0   
6      1009    AL    Blount County                              1.0   
7      1011    AL   Bullock County                              6.0   
8      1013    AL    Butler County                              6.0   
9      1015    AL   Calhoun County                              3.0   
10     1017    AL  Chambers County                              6.0   
11     1019    AL  Cherokee County                              8.0   
12     1021    AL   Chilton County                              1.0   
13     1023    AL   Choctaw County                              9.0   
14     1025    AL    Clarke County                              7.0   

    Rural-urban_Continuum Code_2013  Urban_Influence_Code_2003  \
0                               NaN                        NaN   
1                               NaN                        NaN   
2                               2.0                        2.0   
3                               3.0                        5.0   
4                               6.0                        6.0   
5                               1.0                        1.0   
6                               1.0                        1.0   
7                               6.0                        6.0   
8                               6.0                        6.0   
9                               3.0                        2.0   
10                              6.0                        5.0   
11                              6.0                        6.0   
12                              1.0                        1.0   
13                              9.0                       10.0   
14                              7.0                       11.0   

    Urban_Influence_Code_2013  Economic_typology_2015 CENSUS_2010_POP  \
0                         NaN                     NaN     308,745,538   
1                         NaN                     NaN       4,779,736   
2                         2.0                     0.0          54,571   
3                         2.0                     5.0         182,265   
4                         6.0                     3.0          27,457   
5                         1.0                     0.0          22,915   
6                         1.0                     0.0          57,322   
7                         6.0                     3.0          10,914   
8                         6.0                     0.0          20,947   
9                         2.0                     4.0         118,572   
10                        5.0                     0.0          34,215   
11                        6.0                     0.0          25,989   
12                        1.0                     0.0          43,643   
13                       10.0                     3.0          13,859   
14                       11.0                     3.0          25,833   

   ESTIMATES_BASE_2010  ... R_DOMESTIC_MIG_2019 R_NET_MIG_2011 R_NET_MIG_2012  \
0          308,758,105  ...                 NaN            NaN            NaN   
1            4,780,125  ...                 1.9            0.6            1.2   
2               54,597  ...                 4.8            6.0           -6.2   
3              182,265  ...                24.0           16.6           17.5   
4               27,455  ...                -5.7            0.3           -6.9   
5               22,915  ...                 1.4           -5.0           -3.8   
6               57,322  ...                 1.0            0.2           -

In [5]:
pop.shape

(3273, 165)

Population Data has 165 columns, some which are not needed, and some that need to be converted into rows.

# Shape DataFrame and Delete Columns

In [6]:
# Remove Rural-urban_Continuum columns
pop = pop[pop.columns.drop(list(pop.filter(regex='Rural-urban_Continuum')))]

In [7]:
pop.columns.tolist()

['FIPStxt',
 'State',
 'Area_Name',
 'Urban_Influence_Code_2003',
 'Urban_Influence_Code_2013',
 'Economic_typology_2015',
 'CENSUS_2010_POP',
 'ESTIMATES_BASE_2010',
 'POP_ESTIMATE_2010',
 'POP_ESTIMATE_2011',
 'POP_ESTIMATE_2012',
 'POP_ESTIMATE_2013',
 'POP_ESTIMATE_2014',
 'POP_ESTIMATE_2015',
 'POP_ESTIMATE_2016',
 'POP_ESTIMATE_2017',
 'POP_ESTIMATE_2018',
 'POP_ESTIMATE_2019',
 'N_POP_CHG_2010',
 'N_POP_CHG_2011',
 'N_POP_CHG_2012',
 'N_POP_CHG_2013',
 'N_POP_CHG_2014',
 'N_POP_CHG_2015',
 'N_POP_CHG_2016',
 'N_POP_CHG_2017',
 'N_POP_CHG_2018',
 'N_POP_CHG_2019',
 'Births_2010',
 'Births_2011',
 'Births_2012',
 'Births_2013',
 'Births_2014',
 'Births_2015',
 'Births_2016',
 'Births_2017',
 'Births_2018',
 'Births_2019',
 'Deaths_2010',
 'Deaths_2011',
 'Deaths_2012',
 'Deaths_2013',
 'Deaths_2014',
 'Deaths_2015',
 'Deaths_2016',
 'Deaths_2017',
 'Deaths_2018',
 'Deaths_2019',
 'NATURAL_INC_2010',
 'NATURAL_INC_2011',
 'NATURAL_INC_2012',
 'NATURAL_INC_2013',
 'NATURAL_INC_2014'

In [8]:
# Remove GQ_ESTIMATES
pop = pop[pop.columns.drop(list(pop.filter(regex='GQ_ESTIMATES')))]

In [9]:
# Remove R_ (rates)
pop = pop[pop.columns.drop(list(pop.filter(regex='R_')))]

In [10]:
# Remove RESIDUAL
pop = pop[pop.columns.drop(list(pop.filter(regex='RESIDUAL')))]

In [11]:
# Remove NET_MIG
pop = pop[pop.columns.drop(list(pop.filter(regex='NET_MIG')))]

In [12]:
# Remove remaining unnecessary columns
pop = pop[pop.columns.drop(['Urban_Influence_Code_2003',
 'Urban_Influence_Code_2013',
 'Economic_typology_2015',
 'CENSUS_2010_POP',
 'ESTIMATES_BASE_2010'])]

In [13]:
pop.shape

(3273, 73)

I need to break the DataFrame into multiple DataFrames in order to use the `.melt` function and then bring them back together into a shape that I want.

In [14]:
#Population Estimate
pop_est_list = ['FIPStxt','State','Area_Name','POP_ESTIMATE_2010',
 'POP_ESTIMATE_2011','POP_ESTIMATE_2012','POP_ESTIMATE_2013',
 'POP_ESTIMATE_2014','POP_ESTIMATE_2015','POP_ESTIMATE_2016',
 'POP_ESTIMATE_2017','POP_ESTIMATE_2018','POP_ESTIMATE_2019']

pop_est = pop.filter(pop_est_list)

In [15]:
pop_est

FIPStxt State                         Area_Name POP_ESTIMATE_2010  \
0           0    US                     United States       309,321,666   
1        1000    AL                           Alabama         4,785,437   
2        1001    AL                    Autauga County            54,773   
3        1003    AL                    Baldwin County           183,112   
4        1005    AL                    Barbour County            27,327   
...       ...   ...                               ...               ...   
3268    72145    PR  Vega Baja Municipio, Puerto Rico            59,562   
3269    72147    PR    Vieques Municipio, Puerto Rico             9,314   
3270    72149    PR   Villalba Municipio, Puerto Rico            26,001   
3271    72151    PR    Yabucoa Municipio, Puerto Rico            37,874   
3272    72153    PR      Yauco Municipio, Puerto Rico            41,828   

     POP_ESTIMATE_2011 POP_ESTIMATE_2012 POP_ESTIMATE_2013 POP_ESTIMATE_2014  \
0          311,556,874       313,830,990       315,993,715       318,301,008   
1            4,799,069         4,815,588         4,830,081         4,841,799   
2               55,227            54,954            54,727            54,893   
3              186,558           190,145           194,885           199,183   
4               27,341            27,169            26,937            26,755   
...                ...               ...               ...               ...   
3268            58,688            57,857            57,063            55,942   
3269             9,247             9,227             9,179             9,081   
3270            25,542            25,135            24,745            24,256   
3271            37,361            36,844            36,377            35,732   
3272            41,036            40,249            39,492            38,535   

     POP_ESTIMATE_2015 POP_ESTIMATE_2016 POP_ESTIMATE_2017 POP_ESTIMATE_2018  \
0          320,635,163       322,941,311       324,985,539       326,687,501   
1            4,852,347         4,863,525         4,874,486         4,887,681   
2               54,864            55,243            55,390            55,533   
3              202,939           207,601           212,521           217,855   
4               26,283            25,806            25,157            24,872   
...                ...               ...               ...               ...   
3268            54,805            53,677            52,318            50,136   
3269             8,965             8,833             8,666             8,362   
3270            23,662            23,091            22,452            21,436   
3271            35,076            34,363            33,519            32,254   
3272            37,585            36,654            35,474            33,853   

     POP_ESTIMATE_2019  
0          328,239,523  
1            4,903,185  
2               55,869  
3              223,234  
4               24,686  
...                ...  
3268            50,023  
3269             8,386  
3270            21,372  
3271            32,282  
3272            33,575  

[3273 rows x 13 columns]

In [16]:
# remove POP_ESTIMATE_ from columns
pop_est.columns = [col.replace('POP_ESTIMATE_','') for col in pop_est.columns]

In [17]:
pop_est = pd.melt(pop_est,id_vars=['FIPStxt','State','Area_Name'],
                 var_name = 'Year',value_name = 'Pop_Est')

In [18]:
#N_POP
n_pop_list = ['FIPStxt','State','Area_Name','N_POP_CHG_2010',
 'N_POP_CHG_2011','N_POP_CHG_2012','N_POP_CHG_2013',
 'N_POP_CHG_2014','N_POP_CHG_2015','N_POP_CHG_2016',
 'N_POP_CHG_2017','N_POP_CHG_2018','N_POP_CHG_2019']
n_pop = pop.filter(n_pop_list)

In [19]:
# remove N_POP_CHG_ from columns
n_pop.columns = [col.replace('N_POP_CHG_','') for col in n_pop.columns]

In [20]:
n_pop = pd.melt(n_pop,id_vars=['FIPStxt','State','Area_Name'],
                 var_name = 'Year',value_name = 'Pop_Change')

In [21]:
#Births
birth_list = ['FIPStxt','State','Area_Name','Births_2010',
 'Births_2011','Births_2012','Births_2013',
 'Births_2014','Births_2015','Births_2016',
 'Births_2017','Births_2018','Births_2019']
birth = pop.filter(birth_list)

In [22]:
# remove Births_ from columns
birth.columns = [col.replace('Births_','') for col in birth.columns]

In [23]:
birth = pd.melt(birth,id_vars=['FIPStxt','State','Area_Name'],
                 var_name = 'Year',value_name = 'Births')

In [24]:
#Deaths
death_list = ['FIPStxt','State','Area_Name','Deaths_2010',
 'Deaths_2011','Deaths_2012','Deaths_2013',
 'Deaths_2014','Deaths_2015','Deaths_2016',
 'Deaths_2017','Deaths_2018','Deaths_2019']
death = pop.filter(death_list)

# remove Deaths_ from columns
death.columns = [col.replace('Deaths_','') for col in death.columns]

# reduce into rows
death = pd.melt(death,id_vars=['FIPStxt','State','Area_Name'],
                 var_name = 'Year',value_name = 'Deaths')

In [25]:
#Natural Increases
nat_inc_list = ['FIPStxt','State','Area_Name','NATURAL_INC_2010',
 'NATURAL_INC_2011','NATURAL_INC_2012','NATURAL_INC_2013',
 'NATURAL_INC_2014','NATURAL_INC_2015','NATURAL_INC_2016',
 'NATURAL_INC_2017','NATURAL_INC_2018','NATURAL_INC_2019']
nat_inc = pop.filter(nat_inc_list)

# remove NATURAL_INC from columns
nat_inc.columns = [col.replace('NATURAL_INC_','') for col in nat_inc.columns]

# reduce into rows
nat_inc = pd.melt(nat_inc,id_vars=['FIPStxt','State','Area_Name'],
                 var_name = 'Year',value_name = 'Nat_Inc')

In [26]:
#International Migrations
int_mig_list = ['FIPStxt','State','Area_Name','INTERNATIONAL_MIG_2010',
 'INTERNATIONAL_MIG_2011','INTERNATIONAL_MIG_2012','INTERNATIONAL_MIG_2013',
 'INTERNATIONAL_MIG_2014','INTERNATIONAL_MIG_2015','INTERNATIONAL_MIG_2016',
 'INTERNATIONAL_MIG_2017','INTERNATIONAL_MIG_2018','INTERNATIONAL_MIG_2019']
int_mig = pop.filter(int_mig_list)

# remove INTERNATIONAL_MIG from columns
int_mig.columns = [col.replace('INTERNATIONAL_MIG_','') for col in int_mig.columns]

# reduce into rows
int_mig = pd.melt(int_mig,id_vars=['FIPStxt','State','Area_Name'],
                 var_name = 'Year',value_name = 'Int_Mig')

In [27]:
#Domestic Migrations
dom_mig_list = ['FIPStxt','State','Area_Name','DOMESTIC_MIG_2010',
 'DOMESTIC_MIG_2011','DOMESTIC_MIG_2012','DOMESTIC_MIG_2013',
 'DOMESTIC_MIG_2014','DOMESTIC_MIG_2015','DOMESTIC_MIG_2016',
 'DOMESTIC_MIG_2017','DOMESTIC_MIG_2018','DOMESTIC_MIG_2019']
dom_mig = pop.filter(dom_mig_list)

# remove DOMESTIC_MIG from columns
dom_mig.columns = [col.replace('DOMESTIC_MIG_','') for col in dom_mig.columns]

# reduce into rows
dom_mig = pd.melt(dom_mig,id_vars=['FIPStxt','State','Area_Name'],
                 var_name = 'Year',value_name = 'Dom_Mig')

In [28]:
pop_df = pd.merge(dom_mig,int_mig,how = 'inner',
                  on=['FIPStxt','State','Area_Name','Year'])

In [29]:
pop_df = pop_df.merge(nat_inc,how = 'inner',
                      on=['FIPStxt','State','Area_Name','Year'])

In [30]:
pop_df = pop_df.merge(death,how = 'inner',
                      on=['FIPStxt','State','Area_Name','Year'])

In [31]:
pop_df = pop_df.merge(birth,how = 'inner',
                      on=['FIPStxt','State','Area_Name','Year'])

In [32]:
pop_df = pop_df.merge(n_pop,how = 'inner',
                      on=['FIPStxt','State','Area_Name','Year'])

In [33]:
pop_df = pop_df.merge(pop_est,how = 'inner',
                      on=['FIPStxt','State','Area_Name','Year'])

In [34]:
pop_df.sort_values(['FIPStxt','Year'],inplace=True)

In [35]:
pop_df

FIPStxt State                     Area_Name  Year Dom_Mig  Int_Mig  \
0            0    US                 United States  2010       0  174,416   
3273         0    US                 United States  2011       0  774,165   
6546         0    US                 United States  2012       0  838,671   
9819         0    US                 United States  2013       0  830,168   
13092        0    US                 United States  2014       0  926,546   
...        ...   ...                           ...   ...     ...      ...   
19637    72153    PR  Yauco Municipio, Puerto Rico  2015     NaN      NaN   
22910    72153    PR  Yauco Municipio, Puerto Rico  2016     NaN      NaN   
26183    72153    PR  Yauco Municipio, Puerto Rico  2017     NaN      NaN   
29456    72153    PR  Yauco Municipio, Puerto Rico  2018     NaN      NaN   
32729    72153    PR  Yauco Municipio, Puerto Rico  2019     NaN      NaN   

         Nat_Inc     Deaths     Births Pop_Change      Pop_Est  
0        389,145    598,691    987,836    563,561  309,321,666  
3273   1,461,043  2,512,442  3,973,485  2,235,208  311,556,874  
6546   1,435,445  2,501,531  3,936,976  2,274,116  313,830,990  
9819   1,332,557  2,608,019  3,940,576  2,162,725  315,993,715  
13092  1,380,747  2,582,448  3,963,195  2,307,293  318,301,008  
...          ...        ...        ...        ...          ...  
19637        NaN        NaN        NaN        NaN       37,585  
22910        NaN        NaN        NaN        NaN       36,654  
26183        NaN        NaN        NaN        NaN       35,474  
29456        NaN        NaN        NaN        NaN       33,853  
32729        NaN        NaN        NaN        NaN       33,575  

[32730 rows x 11 columns]

In [36]:
pop_df.to_pickle(os.path.join(path,'Data','Prepared Data','population_organized.pkl'))